### ISSUE TITLE
'weight' must be 2-D

### ISSUE THREAD

**Issue by 21JayChou** [modelscope/ms-swift#4656]
以下是我的训练脚本：

> CUDA_VISIBLE_DEVICES=0,1,2,3 \
> NPROC_PER_NODE=4 \
> MAX_PIXELS=611520 \
> swift rlhf \
>     --rlhf_type grpo \
>     --model models/qwen2.5vl-sft-v2 \
>     --deepspeed zero3_offload \
>     --external_plugins plugin.py \
>     --reward_funcs semantic_reward \
>     --gradient_checkpointing true \
>     --train_type full \
>     --torch_dtype bfloat16 \
>     --dataset dataset_grpo.jsonl \
>     --split_dataset_ratio 0.02 \
>     --max_length 8192 \
>     --max_completion_length 1024 \
>     --num_train_epochs 1 \
>     --per_device_train_batch_size 4 \
>     --per_device_eval_batch_size 4 \
>     --learning_rate 1e-6 \
>     --gradient_accumulation_steps 2 \
>     --eval_steps 10 \
>     --save_steps 50 \
>     --save_total_limit 2 \
>     --logging_steps 1 \
>     --output_dir models \
>     --warmup_ratio 0.05 \
>     --dataloader_num_workers 4 \
>     --num_generations 4 \
>     --temperature 1.0 \
>     --log_completions true \
>     --beta 0.02 \
>     --num_iterations 2 \
>     --loss_type bnpo \
>     --epsilon_high 0.28 \
>     --dynamic_sample true \
>     --overlong_filter true \
>     --max_resample_times 3
>     

其中semantic_reward是我自定义的奖励函数，其中用到了本地部署的qwen3-8b模型，具体实现如下：
```
class SemanticReward(ORM):
    _lock_file = "semantic_reward_class_init.lock"
    def __init__(self):
        model_name = "models/Qwen3-8B"
        # load the tokenizer and the model
        self.tokenizer = AutoTokenizer.from_pretrained(model_name)
        self.model = AutoModelForCausalLM.from_pretrained(
        model_name,
        torch_dtype="bfloat16",
        )
        print(f'reward model:{self.model}')
        
    def chat_local_model(self, system_prompt=None, user_prompt=None):
        messages = [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_prompt}
        ]
        text = self.tokenizer.apply_chat_template(
            messages,
            tokenize=False,
            add_generation_prompt=True,
            enable_thinking=True # Switches between thinking and non-thinking modes. Default is True.
        )
        model_inputs = self.tokenizer([text], return_tensors="pt").to(self.model.device)
    
        # conduct text completion
        generated_ids = self.model.generate(
            **model_inputs,
            max_new_tokens=32768,
            temperature=0.1
        )
        output_ids = generated_ids[0][len(model_inputs.input_ids[0]):].tolist() 
        # parsing thinking content
        try:
            # rindex finding 151668 (</think>)
            index = len(output_ids) - output_ids[::-1].index(151668)
        except ValueError:
            index = 0
        
        thinking_content = self.tokenizer.decode(output_ids[:index], skip_special_tokens=True).strip("\n")
        content = self.tokenizer.decode(output_ids[index:], skip_special_tokens=True).strip("\n")
        
        # print("thinking content:", thinking_content)
        print("content:", content)
        return content
    
    def llm_text_similary(self, text1, text2):
        system_content = """
        有两个关于APP UI展示问题的描述，计算两个描述之间的语义相似度，分数范围0.0-1.0，仅输出分数。 
        """
        user_content = f"""
        === 实际描述 ===
        {text1}
        === 预测描述 ===
        {text2}
        """
        for i in range(3):
            try:
                ans = self.chat_local_model(system_prompt=system_content, user_prompt=user_content)
                score = float(ans.strip())
                return score
            except Exception as e:
                print(traceback.format_exc())
                print(f'第{i}次生成出错:{e}')
        
    def __call__(self, completions, solution, **kwargs) -> List[float]:
        rewards = []
        for content, sol in zip(completions, solution):
            reward = 0.0
            if reward == 0.0:
                try:
                    sol_match = re.search(r'<answer>(.*?)</answer>', sol.replace(" ", "").replace("\n", ""))
                    ground_truth = sol_match.group(1).strip() if sol_match else sol.strip()
                    print(f'ground_truth:{ground_truth}')
                    content_match = re.search(r'<answer>(.*?)</answer>', content.replace(" ", "").replace("\n", ""))
                    student_answer = content_match.group(1).strip() if content_match else content.strip()
                    print(f'student_answer:{student_answer}')
                    content_match = re.search(r'<think>(.*?)</think>', content.replace(" ", "").replace("\n", ""))
                    student_think = content_match.group(1).strip() if content_match else content.strip()
                    
                    score = self.llm_text_similary(ground_truth, student_answer)
                    if score >= 0.8:
                        reward = 1.0
                    elif score <= 0.3:
                        reward = 0.0
                    else:
                        reward = score
                    reward = round(reward, 2)
                except Exception:
                    pass
            rewards.append(reward)
        return rewards
```
在训练过程中奖励函数在使用qwen3-8B进行推理时，就会出现'weight' must be 2-D的报错：

<img width="632" alt="Image" src="https://github.com/user-attachments/assets/a5c06ba0-b19d-4734-ba81-588a50086200" />

**Comment by hjh0119** [comment]
maybe related issue https://github.com/modelscope/ms-swift/issues/4580

**Comment by 21JayChou** [comment]
我改成使用reward_model后,reward_model似乎并没有输出，脚本如下：

> CUDA_VISIBLE_DEVICES=0,1,2,3 \
> NPROC_PER_NODE=4 \
> MAX_PIXELS=611520 \
> swift rlhf \
>     --rlhf_type grpo \
>     --model models/qwen2.5vl-sft-v2 \
>     --deepspeed zero3 \
>     --use_vllm true \
>     --vllm_mode colocate \
>     --vllm_max_model_len 8192 \
>     --vllm_gpu_memory_utilization 0.5 \
>     --offload_optimizer true \
>     --offload_model true \
>     --gc_collect_after_offload true \
>     --external_plugins plugin.py \
>     --reward_model_plugin semantic_reward \
>     --reward_model models/Qwen3-8B \
>     --gradient_checkpointing true \
>     --train_type full \
>     --torch_dtype bfloat16 \
>     --dataset dataset_grpo.jsonl \
>     --split_dataset_ratio 0.02 \
>     --max_length 8192 \
>     --max_completion_length 1024 \
>     --num_train_epochs 1 \
>     --per_device_train_batch_size 4 \
>     --per_device_eval_batch_size 4 \
>     --learning_rate 1e-6 \
>     --gradient_accumulation_steps 2 \
>     --eval_steps 10 \
>     --save_steps 50 \
>     --save_total_limit 2 \
>     --logging_steps 1 \
>     --output_dir models \
>     --warmup_ratio 0.05 \
>     --dataloader_num_workers 4 \
>     --num_generations 4 \
>     --temperature 1.0 \
>     --log_completions true \
>     --beta 0.02 \
>     --num_iterations 2 \
>     --loss_type bnpo \
>     --epsilon_high 0.28 \
>     --dynamic_sample true \
>     --overlong_filter true \
>     --max_resample_times 3

    reward_model_pluging定义如下：

```
class SematnticRewardPlugin(DefaultRMPlugin):
    # https://arxiv.org/abs/2505.17667
    # NOTE: you should customize the verified reward function, you can refer to
    # https://github.com/Tongyi-Zhiwen/QwenLong-L1/tree/main/verl/verl/utils/reward_score
    # hf_dataset: https://huggingface.co/datasets/Tongyi-Zhiwen/DocQA-RL-1.6K/viewer/default/train
    # ms_dataset: https://modelscope.cn/datasets/iic/DocQA-RL-1.6K
    def __init__(self, model, template, accuracy_orm=None):
        super().__init__(model, template)
        # initilize PTEngine to infer
        self.engine = PtEngine.from_model_template(self.model, self.template, max_batch_size=0)  # 0: no limit
        self.request_config = RequestConfig(temperature=0.1)  # customise your request config here
        self.system = """
        有两个关于APP UI展示问题的描述，计算两个描述之间的语义相似度，分数范围0.0-1.0，仅输出分数。 
        """ 
        self.user = """
        === 实际描述 ===
        {text1}
        === 预测描述 ===
        {text2}
        """
        self.accuracy_orm = accuracy_orm

    def __call__(self, inputs):
        completions = [example['messages'][-1]['content'] for example in inputs]
        ground_truths = [example['solution'] for example in inputs]
        rm_inputs = self.prepare_rm_inputs(inputs, completions, ground_truths)

        results = self.engine.infer(rm_inputs, self.request_config, use_tqdm=False)
        llm_rewards = self.compute_rewards(results)

        if self.accuracy_orm:
            verified_rewards = self.accuracy_orm(completions, ground_truths)
        else:
            verified_rewards = [0.0] * len(llm_rewards)

        rewards = [max(r1, r2) for r1, r2 in zip(llm_rewards, verified_rewards)]
        return torch.tensor(rewards, dtype=torch.float32)

    def prepare_rm_inputs(self, inputs: List[Dict], completions, ground_truths) -> List[Dict]:
        rm_inputs = []
        for infer_request, completion, ground_truth in zip(inputs, completions, ground_truths):
            # Deep copy to prevent modification of original input
            content_match = re.search(r'<answer>(.*?)</answer>', completion.replace(" ", "").replace("\n", ""))
            completion = content_match.group(1).strip() if content_match else content.strip()
            print(f"completion:{completion}")
            print(f"ground_truth:{ground_truth}")
            rm_infer_request = deepcopy(infer_request)
            user_prompt = self.user.format(
                text1=ground_truth, text2=completion)

            # Construct new messages tailored for the reward model
            rm_messages = [
            {"role": "system", "content":self.system},
            {"role": "user", "content": user_prompt}
            ]

            # Update the messages in the reward infer request
            rm_infer_request['messages'] = rm_messages
            rm_inputs.append(rm_infer_request)
        return rm_inputs

    @staticmethod
    def extract_reward(model_output: str) -> float:
        reward = 0.0
        try:
            score = float(model_output.strip())
            print(f"score:{score}")
            if score >= 0.8:
                reward = 1.0
            elif score <= 0.3:
                reward = 0.0
            else:
                reward = score
            reward = round(reward, 2)
            return reward
        except:
            logger.warning("Unable to extract reward score from the model's output, setting reward to 0")
            return 0.0  # Or raise ValueError("Format incorrect")

    def compute_rewards(self, results: List[ChatCompletionResponse]) -> List[float]:
        """
        Compute average reward scores from the reward model's outputs.

        Args:
            results (List[ChatCompletionResponse]): A list of results from the reward model.

        Returns:
            List[float]: A list of average reward scores.
        """
        rewards = []
        for idx, output in enumerate(results):
            try:
                cur_rewards = []
                for choice in output.choices:
                    response = choice.message.content
                    reward = self.extract_reward(response)
                    cur_rewards.append(reward)
                cur_rewards = [r for r in cur_rewards if r is not None]
                if cur_rewards:
                    average_reward = sum(cur_rewards) / len(cur_rewards)
                else:
                    average_reward = 0.0
                    logger.warning('No valid rewards extracted. Assigning reward score of 0.0.')

                rewards.append(average_reward)
            except Exception as e:
                logger.error(f'Error computing reward: {e}')
                rewards.append(0.0)  # Assign default reward score on failure
        return rewards
    
rm_plugins['semantic_reward'] = SematnticRewardPlugin
```
    实际运行中会有：
    

> [WARNING:swift] Unable to extract reward score from the model's output, setting reward to 0
> [WARNING:swift] Unable to extract reward score from the model's output, setting reward to 0
> [WARNING:swift] Unable to extract reward score from the model's output, setting reward to 0
> [WARNING:swift] Unable to extract reward score from the model's output, setting reward to 0
> [WARNING:swift] Unable to extract reward score from the model's output, setting reward to 0
> [WARNING:swift] Unable to extract reward score from the model's output, setting reward to 0
> [WARNING:swift] Unable to extract reward score from the model's output, setting reward to 0
> [WARNING:swift] Unable to extract reward score from the model's output, setting reward to 0

**Comment by hjh0119** [comment]
You can debug to see why the exception occurred; it might be due to an error during the step where the model output is converted to float.

**Comment by 21JayChou** [comment]
> You can debug to see why the exception occurred; it might be due to an error during the step where the model output is converted to float.

确实是这里的问题，模型输出包含think过程了，没有过滤这一段，感谢大佬！！
但是rm的输出时间都很长，要十几分钟，不知道是什么情况

**Comment by hjh0119** [comment]
PTEngine has a slower generation speed, especially for larger models and longer texts, and since the output of Qwen3 is generally quite long (typically around 4000 tokens), the time taken is relatively long.

As an alternative, you can deploy using vLLM or other efficient inference frameworks and then call the service in the reward_funcs.

**Comment by 21JayChou** [comment]
> PTEngine has a slower generation speed, especially for larger models and longer texts, and since the output of Qwen3 is generally quite long (typically around 4000 tokens), the time taken is relatively long.
> 
> As an alternative, you can deploy using vLLM or other efficient inference frameworks and then call the service in the reward_funcs.

目前服务器不支持我把模型部署成一个service，这样其他进程是无法访问的，还是非常感谢！

**Comment by 21JayChou** [comment]
又有新的问题，grpo过程中显存一直增长，直到OOM，有什么参数可以调整么
